### PR TITLE
Create new configuration caching compatible enterprise plugin contact point

### PR DESCRIPTION
**Pull request by ldaley**
This change also moves the contact point out to a new `enterprise` module.

This PR is a bit noisy. Half way through adding the new contact point I decided to extract the new and old to a new module, which was more disruptive than I anticipated.

https://github.com/gradle/gradle/tree/ldaley/scan-config-cache/subprojects/enterprise/src/main/java/org/gradle/internal/scan is the old entry point, and is functionally unchanged. It will be used by plugin 3.3 and earlier.

https://github.com/gradle/gradle/tree/ldaley/scan-config-cache/subprojects/enterprise/src/main/java/org/gradle/internal/enterprise is the new contact point which is configuration cache compatible.

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle help

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test QF

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

File: subprojects/core/src/integTest/groovy/org/gradle/internal/scopeids/ScopeIdsIntegrationTest.groovy [hunk 1]
```diff
@@ -155,20 +154,4 @@ class ScopeIdsIntegrationTest extends AbstractIntegrationSpec {
         scopeIds.userIds.unique().size() == 2
     }
 
-    def "exposes scans view of scope IDs"() {
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/core/src/integTest/groovy/org/gradle/internal/scopeids/ScopeIdsIntegrationTest.groovy:158]: This is tested elsewhere in the “legacy” package of the new module. This was duplicative.

File: subprojects/core/src/main/java/org/gradle/invocation/DefaultGradle.java [hunk 2]
```diff
@@ -105,8 +113,29 @@ public void projectsLoaded(Gradle gradle) {
...
+        registerMissingBuildScanPluginWarningListenerIfRootBuild(parent, startParameter);
+    }
+
+    /**
+     * Causes a warning to be emitted if the --scan was used but the plugin didn't show up.
+     *
+     * This should never happen due to the auto apply behavior.
+     * It's only here as a kind of safeguard or fallback.
+     */
+    private void registerMissingBuildScanPluginWarningListenerIfRootBuild(GradleInternal parent, StartParameter startParameter) {
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/core/src/main/java/org/gradle/invocation/DefaultGradle.java:125]: Good call. Will do.

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Pull request merged**